triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/UnfoldState.scala

```diff
@@ -86,45 +112,94 @@ private[trigger] object UnfoldState {
     */
   def flatMapConcatStates[T, A, B](zero: T)(
       f: (T, A) => UnfoldState[T, B]): Flow[A, T \/ B, NotUsed] =
-    Flow[A].statefulMapConcat { () =>
-      var t = zero
-      // statefulMapConcat only uses 'iterator'.  We preserve the Iterable's
-      // immutability by making one strict reference to the 't' var at creation
-      // time, meaning any later 'iterator' call uses the same start state, no matter
-      // whether the above 't' has been updated
-      a =>
-        new Iterable[T \/ B] {
-          private[this] val bs = f(t, a)
-          import bs.step
-          override def iterator = new Iterator[T \/ B] {
-            private[this] var last: Option[T \/ (B, bs.S)] = {
-              val fst = step(bs.init)
-              fst fold (newT => t = newT, _ => ())
-              Some(fst)
-            }
+    Flow[A].statefulMapConcat(() => mkMapConcatFun(zero, f))
 
-            // this stream is "odd", i.e. we are always evaluating 1 step ahead
-            // of what the client sees.  We could improve laziness by making it
-            // "even", but it would be a little trickier, as `hasNext` would have
-            // a forcing side-effect
-            override def hasNext() = last.isDefined
-
-            override def next() =
-              last match {
-                case Some(\/-((b, s))) =>
-                  val next = step(s)
-                  // The assumption here is that statefulMapConcat's implementation
-                  // will always read iterator to end before invoking on the next A
-                  next fold (newT => t = newT, _ => ())
-                  last = Some(next)
-                  \/-(b)
-                case Some(et @ -\/(_)) =>
-                  last = None
-                  et
-                case None =>
-                  throw new IllegalStateException("iterator read past end")
-              }
+  /** Accept 1 initial state on in1, fold over in2 elements, emitting the output
+    * elements on out1, and each result state after each result of `f` is unfolded
+    * on out2.
+    */
+  def flatMapConcatNode[T, A, B](
```

Comment:
I find the use of `BidiShape` a bit confusing. The initial state and the messages are both inputs on the “same” side in my mental model. I realize Akka doesn’t care about this and `BidiShape` is just something with two inputs and two outputs without any requirements on the connection between those but maybe it would still be useful to define something custom to avoid confusion?

The single element initial state stream also feels slightly weird to me but I don’t have a better solution for that unfortunately.

Reply:
> I realize Akka doesn’t care about this and `BidiShape` is just something with two inputs and two outputs without any requirements on the connection between those but maybe it would still be useful to define something custom to avoid confusion?

Sure. You think type alias, or something like `SourceShape2` (also in this PR)? Words, more code...

I don't like redefining shapes, but I think adding methods implicitly to replace names like `out1` might be useful. What do you think of that?

> The single element initial state stream also feels slightly weird to me but I don’t have a better solution for that unfortunately.

In general variants of `Inlet` and `Outlet` limited to exactly one element would be a useful abstraction in several ways, but not something we can add ourselves, I think. We cope with this lack of distinction in JSON API's streams, too. It is unfortunate.

Reply:
I probably would have defined something like `SourceShape2` but I’m also happy with just a type alias or an implicit method. Whatever you prefer.

Reply:
eb276b36d6